File: unity/Pancing/Assets/Pancing/Sim/Data/Spots.cs
using System;
using System.Collections.Generic;

namespace Pancing.Sim
{
    /// <summary>A snag point. A fish that reaches one and stays there breaks you off.</summary>
    public struct SpotStructure
    {
        /// <summary>Normalised position across the castable water: U from shore, V lateral.</summary>
        public double U, V, R;
        public string Kind;
    }

    /// <summary>Colours driving the procedural environment generator.</summary>
    public sealed class SpotPalette
    {
        public string Water, Deep, Shallow, Foam, Sand, Grass, Trees;
        public string[] Sky = System.Array.Empty<string>();
    }

    public sealed class Spot
    {
        public string Id, Name, Tagline;
        public int Level;
        public double EntryFee, MaxDepth, WaterClarity, Current, WindBase, SnagDensity;
        /// <summary>
        /// Species ids available here, with a spot-local weight tweak.
        ///
        /// A LIST, not a dictionary, and the order is the order they appear in
        /// spots.json. The catch table walks this order to resolve a weighted
        /// draw, so reordering it changes which fish a given random number picks.
        /// </summary>
        public List<KeyValuePair<string, double>> Pool = new List<KeyValuePair<string, double>>();
        public SpotPalette Palette = new SpotPalette();
        public SpotStructure[] Structure = System.Array.Empty<SpotStructure>();

        /// <summary>
        /// Normalised depth 0..1 across the castable water. The renderer shapes the
        /// lake bed with this same function, so what you see is genuinely what you
        /// are fishing — the water cannot be deep where the ground is high.
        /// </summary>
        public double DepthAt(double u, double v) => SpotShapes.DepthAt(Id, u, v);
    }

    public sealed class TimePhase
    {
        public string Id, Label;
        public double From, To, Sun, Warm;
    }

    public sealed class WeatherSpec
    {
        public string Id, Label;
        public double Chance, Wind, Chop, Light, Rain;
    }

    /// <summary>
    /// Lake-bed shapes.
    ///
    /// This is the one piece of the data tables that is genuinely code rather than
    /// data — `depthAt` is a function of two variables, and JSON has no way to say
    /// "a Gaussian channel offset to one side". Rather than invent an expression
    /// language for three formulas, both engines carry the same three formulas,
    /// keyed by spot id, and shared/parity pins them against each other.
    ///
    /// The JavaScript originals live in web/src/data/spots.js. Keep them in step.
    /// </summary>
    public static class SpotShapes
    {
        public static double DepthAt(string spotId, double u, double v)
        {
            switch (spotId)
            {
                case "kolam":
                {
                    // Gentle dish: deepest in the middle of the pond, shelving to
                    // the banks.
                    double r = MathUtil.Hypot(u - 0.55, v * 0.9);
                    return MathUtil.Clamp01(1.05 - r * 1.5 + Math.Sin(v * 7.0) * 0.03);
                }
                case "sungai":
                {
                    // Channel: a scoured trough offset to one side, shallow gravel
                    // bar opposite.
                    double channel = Math.Exp(-Math.Pow((v + 0.18) / 0.34, 2));
                    return MathUtil.Clamp01(0.18 + u * 0.35 + channel * 0.62 - Math.Abs(v) * 0.15);
                }
                case "tasik":
                {
                    // Steep drop-off close to shore, then a deep flat basin.
                    double shelf = MathUtil.SmoothStep(0.06, 0.30, u);
                    return MathUtil.Clamp01(shelf * 0.92 + u * 0.10 - Math.Abs(v) * 0.06);
                }
                default:
                    // An unknown spot should still be fishable, not a flat zero that
                    // puts every lure on dry land.
                    return MathUtil.Clamp01(0.25 + u * 0.55 - Math.Abs(v) * 0.10);
            }
        }
    }

    public sealed class SpotDb
    {
        public readonly List<Spot> All = new List<Spot>();
        public readonly Dictionary<string, Spot> ById = new Dictionary<string, Spot>();
        public readonly List<TimePhase> Phases = new List<TimePhase>();
        public readonly List<WeatherSpec> Weathers = new List<WeatherSpec>();
        public readonly Dictionary<string, WeatherSpec> WeatherById = new Dictionary<string, WeatherSpec>();

        public Spot Get(string id) =>
            id != null && ById.TryGetValue(id, out var s) ? s : (All.Count > 0 ? All[0] : null);

        public WeatherSpec GetWeather(string id) =>
            id != null && WeatherById.TryGetValue(id, out var w) ? w : (Weathers.Count > 0 ? Weathers[0] : null);

        /// <summary>
        /// The phase containing this hour. Night wraps past midnight, so a phase
        /// whose `to` is less than its `from` matches the union of both ends.
        /// </summary>
        public TimePhase PhaseForHour(double hour)
        {
            double h = ((hour % 24) + 24) % 24;
            foreach (var p in Phases)
            {
                bool inside = p.From < p.To
                    ? (h >= p.From && h < p.To)
                    : (h >= p.From || h < p.To);
                if (inside) return p;
            }
            return Phases.Count > 1 ? Phases[1] : (Phases.Count > 0 ? Phases[0] : null);
        }

        public static SpotDb Load(string json)
        {
            var root = Json.Parse(json);
            var db = new SpotDb();

            foreach (var s in root["spots"].Items)
            {
                var structures = new List<SpotStructure>();
                foreach (var st in s["structure"].Items)
                {
                    structures.Add(new SpotStructure
                    {
                        U = st["u"].AsDouble(),
                        V = st["v"].AsDouble(),
                        R = st["r"].AsDouble(),
                        Kind = st["kind"].AsString("timber"),
                    });
                }

                var pal = s["palette"];
                var spot = new Spot
                {
                    Id = s["id"].AsString(),
                    Name = s["name"].AsString(),
                    Tagline = s["tagline"].AsString(""),
                    Level = s["level"].AsInt(1),
                    EntryFee = s["entryFee"].AsDouble(),
                    MaxDepth = s["maxDepth"].AsDouble(3),
                    WaterClarity = s["waterClarity"].AsDouble(0.5),
                    Current = s["current"].AsDouble(),
                    WindBase = s["windBase"].AsDouble(0.2),
                    SnagDensity = s["snagDensity"].AsDouble(0.2),
                    Pool = s["pool"].AsNumberList(),
                    Structure = structures.ToArray(),
                    Palette = new SpotPalette
                    {
                        Water = pal["water"].AsString("#3f6b5e"),
                        Deep = pal["deep"].AsString("#12302c"),
                        Shallow = pal["shallow"].AsString("#7fae94"),
                        Foam = pal["foam"].AsString("#e8f2ea"),
                        Sand = pal["sand"].AsString("#8d7f5e"),
                        Grass = pal["grass"].AsString("#5c6f43"),
                        Trees = pal["trees"].AsString("#3f5233"),
                        Sky = pal["sky"].AsStringArray(),
                    },
                };

                if (string.IsNullOrEmpty(spot.Id)) continue;
                db.All.Add(spot);
                db.ById[spot.Id] = spot;
            }

            foreach (var p in root["timePhases"].Items)
            {
                db.Phases.Add(new TimePhase
                {
                    Id = p["id"].AsString(),
                    Label = p["label"].AsString(),
                    From = p["from"].AsDouble(),
                    To = p["to"].AsDouble(),
                    Sun = p["sun"].AsDouble(),
                    Warm = p["warm"].AsDouble(),
                });
            }

            foreach (var w in root["weather"].Items)
            {
                var spec = new WeatherSpec
                {
                    Id = w["id"].AsString(),
                    Label = w["label"].AsString(),
                    Chance = w["chance"].AsDouble(),
                    Wind = w["wind"].AsDouble(1),
                    Chop = w["chop"].AsDouble(1),
                    Light = w["light"].AsDouble(1),
                    Rain = w["rain"].AsDouble(),
                };
                db.Weathers.Add(spec);
                db.WeatherById[spec.Id] = spec;
            }

            return db;
        }
    }
}

File: web/src/data/spots.js
/**
 * Fishing locations.
 *
 *  depthAt(u, v)   returns normalised depth 0..1 across the castable water,
 *                  where u is distance from shore and v is lateral offset.
 *                  The renderer uses the same function to shape the lake bed,
 *                  so what you see is genuinely what you are fishing.
 *  structure       snag points; a fish that reaches one breaks you off.
 *  pool            species ids available here, plus a spot-local weight tweak.
 *  palette         drives the procedural environment generator.
 */

export const SPOTS = [
  {
    id: 'kolam', name: 'Kolam Kampung', tagline: 'Air tenang, ikan lapar, tiada tekanan.',
    level: 1, entryFee: 0, maxDepth: 3.2, waterClarity: 0.55, current: 0.05,
    windBase: 0.22, snagDensity: 0.18,
    pool: {
      tilapia: 1.4, keli: 1.2, lampam: 1.1, puyu: 1.3, haruan: 0.5,
      boot: 1.0, tin: 1.0, plastik: 0.8, ranting: 0.9,
    },
    palette: {
      water: '#3f6b5e', deep: '#12302c', shallow: '#7fae94', foam: '#e8f2ea',
      sand: '#8d7f5e', grass: '#5c6f43', sky: ['#8fb8d8', '#dfe9ee'],
      trees: '#3f5233',
    },
    depthAt: (u, v) => {
      // Gentle dish: deepest in the middle of the pond, shelving to the banks.
      const r = Math.hypot(u - 0.55, v * 0.9);
      return clamp01(1.05 - r * 1.5 + Math.sin(v * 7.0) * 0.03);
    },
    structure: [
      { u: 0.32, v: -0.45, r: 0.10, kind: 'reeds' },
      { u: 0.58, v: 0.38, r: 0.12, kind: 'timber' },
      { u: 0.78, v: -0.12, r: 0.09, kind: 'reeds' },
    ],
  },
  {
    id: 'sungai', name: 'Sungai Berbatu', tagline: 'Arus deras. Ikan kuat, batuan tajam.',
    level: 4, entryFee: 40, maxDepth: 5.0, waterClarity: 0.78, current: 0.55,
    windBase: 0.35, snagDensity: 0.52,
    pool: {
      sebarau: 1.6, baung: 1.3, lampam: 0.9, jelawat: 1.2, haruan: 0.8,
      keli: 0.7, udang_galah: 1.1, kelah: 0.6,
      ranting: 1.4, tin: 0.8, boot: 0.6,
    },
    palette: {
      water: '#2f6f78', deep: '#0d2b31', shallow: '#79b6b2', foam: '#f2f7f6',
      sand: '#9a927c', grass: '#4c6b3c', sky: ['#7fa9cc', '#e6eef2'],
      trees: '#33472c',
    },
    depthAt: (u, v) => {
      // Channel: a scoured trough offset to one side, shallow gravel bar opposite.
      const channel = Math.exp(-Math.pow((v + 0.18) / 0.34, 2));
      return clamp01(0.18 + u * 0.35 + channel * 0.62 - Math.abs(v) * 0.15);
    },
    structure: [
      { u: 0.24, v: 0.40, r: 0.11, kind: 'rock' },
      { u: 0.46, v: -0.52, r: 0.13, kind: 'rock' },
      { u: 0.66, v: 0.14, r: 0.10, kind: 'timber' },
      { u: 0.86, v: -0.34, r: 0.12, kind: 'rock' },
    ],
  },
  {
    id: 'tasik', name: 'Tasik Dalam', tagline: 'Air hitam. Sesuatu yang besar tinggal di bawah.',
    level: 8, entryFee: 180, maxDepth: 9.5, waterClarity: 0.32, current: 0.12,
    windBase: 0.48, snagDensity: 0.36,
    pool: {
      toman: 1.5, patin: 1.4, belida: 1.3, haruan: 1.0, baung: 1.1,
      jelawat: 0.9, keli: 0.6, kelah: 0.8, udang_galah: 0.7,
      plastik: 1.1, ranting: 1.0, boot: 0.7,
    },
    palette: {
      water: '#25454f', deep: '#07171d', shallow: '#5a8894', foam: '#dfeaee',
      sand: '#6f6552', grass: '#3d5334', sky: ['#5d7f9e', '#c9d8e0'],
      trees: '#26361f',
    },
    depthAt: (u, v) => {
      // Steep drop-off close to shore, then a deep flat basin.
      const shelf = smoothstep(0.06, 0.30, u);
      return clamp01(shelf * 0.92 + u * 0.10 - Math.abs(v) * 0.06);
    },
    structure: [
      { u: 0.18, v: -0.30, r: 0.10, kind: 'timber' },
      { u: 0.40, v: 0.48, r: 0.14, kind: 'timber' },
      { u: 0.72, v: -0.20, r: 0.12, kind: 'weed' },
      { u: 0.90, v: 0.30, r: 0.13, kind: 'weed' },
    ],
  },
];

export const SPOTS_BY_ID = Object.fromEntries(SPOTS.map((s) => [s.id, s]));

/* --- time & weather ------------------------------------------------------- */

export const TIME_PHASES = [
  { id: 'dawn',  label: 'Subuh',  from: 5.0,  to: 8.0,  sun: 0.18, warm: 0.95 },
  { id: 'day',   label: 'Siang',  from: 8.0,  to: 17.0, sun: 1.00, warm: 0.25 },
  { id: 'dusk',  label: 'Senja',  from: 17.0, to: 19.5, sun: 0.22, warm: 1.00 },
  { id: 'night', label: 'Malam',  from: 19.5, to: 5.0,  sun: 0.05, warm: 0.10 },
];

export function phaseForHour(hour) {
  const h = ((hour % 24) + 24) % 24;
  for (const p of TIME_PHASES) {
    if (p.from < p.to ? h >= p.from && h < p.to : h >= p.from || h < p.to) return p;
  }
  return TIME_PHASES[1];
}

export const WEATHER = [
  { id: 'clear',  label: 'Cerah',   chance: 0.40, wind: 0.6, chop: 0.5, light: 1.00, rain: 0.0 },
  { id: 'cloudy', label: 'Mendung', chance: 0.32, wind: 0.9, chop: 0.8, light: 0.72, rain: 0.0 },
  { id: 'rain',   label: 'Hujan',   chance: 0.20, wind: 1.2, chop: 1.2, light: 0.52, rain: 0.6 },
  { id: 'storm',  label: 'Ribut',   chance: 0.08, wind: 2.0, chop: 2.0, light: 0.35, rain: 1.0 },
];

export const WEATHER_BY_ID = Object.fromEntries(WEATHER.map((w) => [w.id, w]));

/* --- small shared helpers ------------------------------------------------- */

function clamp01(x) { return x < 0 ? 0 : x > 1 ? 1 : x; }

function smoothstep(a, b, x) {
  const t = clamp01((x - a) / (b - a));
  return t * t * (3 - 2 * t);
}
